Demo: some CGI (Python and C) and a lighttpd.conf file for a hands-on example

sebas88 opened the pull request:
It's a simple demo for showing how to create the Cookie, since there is
not much info out there. I took the plone Python implementation as a
base but it needs modifications in order to work. The C-code is an
imitation of the mod_authn_tkt.c code since that's the one that will
actually validate the cookies.

Commit:
Demo: some CGI (Python and C) and a lighttpd.conf file for a hands-on example

It's a simple demo for showing how to create the Cookie, since there is
not much info out there. I took the plone Python implementation as a
base but it needs modifications in order to work. The C-code is an
imitation of the mod_authn_tkt.c code since that's the one that will
actually validate the cookies.

gstrauss commented on demo/lighttpd.conf:
```diff
@@ -0,0 +1,55 @@
+
+index-file.names = ( "index.html" )
+
+server.tag = " " # Do not append headers
```
not necessary; please remove from demo example

gstrauss commented on demo/login.cgi:
```diff
@@ -0,0 +1,106 @@
+- The IP must be encoded as a string (plone was using 4-bytes). The reason mod_authn_tkt
+does this is because an IP address might be IPv4 or it might be IPv6. An IPv6 address
+does not fit into 4 packed bytes (which is the size of an IPv4 address)
+- The timestamp must be encoded in little-endian (plone was using network/big-endian).
```
That might be a bug.  mod_authn_tkt converts timestamp to big-endian with `htonl()` before hashing the timestamp.  However, it also uses `htonl()` in `refresh_cookie()`, which would undo the big-endian to little-endian on little-endian systems.  I'll review the code later today, but this looks like a bug.  For consistency, it should be big-endian throughout.  Thanks for finding this!

gstrauss replied:
With the following patch, the timestamps should be big-endian, including for ticket refresh.  (I have not yet pushed this patch)
```
--- a/mod_authn_tkt.c
+++ b/mod_authn_tkt.c
@@ -246,6 +246,7 @@ static int parse_ticket(authn_tkt *parsed)/*{{{*/
                                       sizeof(parsed->timestamp))) {
         return 0; /*(invalid hex encoding in timestamp)*/
     }
+    parsed->timestamp = htonl(parsed->timestamp);
 
     buffer_copy_string_len(&parsed->uid,
                            tkt + digest_hexlen + TIMESTAMP_HEXLEN,
```

sebas88 replied:
Oh, then the C implementation must be changed as well. Glad to be helpful anyway.
Sligthly off-topic; since it's my first Pull-Request in github, how do I do this after you push the bugfix to master? Rebase on original repo and then (force) push my fork? Or merge your repo and then commit the changes?

gstrauss replied:
If you force push to your branch, it will update this pull request.  I have not yet merged any changes.

gstrauss committed:
use uint32_t to store time through year 2106

use unsigned 32-bit value to store time through year 2106
(as long as mod_authn_tkt is compiled 64-bit with 64-bit time_t)

Also fixes one instance missing endian conversion of timestamp,
which could affect cookie validation and refresh.
(thx sebas88)

x-ref:
  https://github.com/gstrauss/lighttpd-mod_authn_tkt/pull/2

gstrauss commented:
I'll push an update to mod_authn_tkt.c momentarily.

In this pull request, the C code demo login_cgi.c needs a small update to use big-endian 4 bytes from time_t (which might be 64-bit)
```
--- a/demo/login_cgi.c
+++ b/demo/login_cgi.c
@@ -22,7 +22,7 @@ typedef struct authn_tkt_struct {
 
 static void ticket_digest_MD5(authn_tkt *parsed, time_t timestamp, const char *secret)/*{{{*/
 {
-    uint32_t ts = ((uint32_t)timestamp); /*(assumes 32-bit time_t)*/
+    uint32_t ts = htonl((uint32_t)timestamp);/*(unsigned 4 bytes from time_t)*/
     MD5_CTX ctx;
 
     /* Generate the initial digest */
```

Note: The python login.cgi does not run for me.  It has errors with concatenating strings and binary.  Please test and fix.

gstrauss commented:
> Sligthly off-topic; since it's my first Pull-Request in github, how do I do this after you push the bugfix to master? Rebase on original repo and then (force) push my fork? Or merge your repo and then commit the changes?

If https://github.com/gstrauss/lighttpd-mod_authn_tkt is labelled `upstream` and your repo is `origin`
`git pull --rebase upstream master`
`git push -f origin master`

Commit:
Demo: removed unnecessary line from lighttpd.conf

Commit:
Demo: fixed timestamp to be big-endian (network order)

This is needed since commit 8a782a198f3ff4fa19461232bed130bd18300e2e
fixed timestamp's endianness to be big-endian (in previous versions it
was machine-dependent)

Commit:
Demo: updated Python CGI to be compatible with Python 3

Also works with Python 2.7

sebas88 commented:
> I'll push an update to mod_authn_tkt.c momentarily.
> 
> In this pull request, the C code demo login_cgi.c needs a small update to use big-endian 4 bytes from time_t (which might be 64-bit)
> 
> ```
> --- a/demo/login_cgi.c
> +++ b/demo/login_cgi.c
> @@ -22,7 +22,7 @@ typedef struct authn_tkt_struct {
>  
>  static void ticket_digest_MD5(authn_tkt *parsed, time_t timestamp, const char *secret)/*{{{*/
>  {
> -    uint32_t ts = ((uint32_t)timestamp); /*(assumes 32-bit time_t)*/
> +    uint32_t ts = htonl((uint32_t)timestamp);/*(unsigned 4 bytes from time_t)*/
>      MD5_CTX ctx;
>  
>      /* Generate the initial digest */
> ```
> 
> Note: The python login.cgi does not run for me. It has errors with concatenating strings and binary. Please test and fix.

The Python implementation was in Python 2.7 (I know, I know... I shouldn't be using Py2.7 in 2021...), I've updated it to work with both versions now

sebas88 commented:
I've made the changes in sepparate commits. If you prefer it to be a single commit with the final demo just say it and I'll squash and force-push the fork again so the PR is condensed in 1 commit.

gstrauss commented on demo/login.cgi:
```diff
@@ -0,0 +1,105 @@
+def createTicket(secret, userid, tokens=(), user_data='', ip='0.0.0.0',
+                 timestamp=None, encoding='utf-8', mod_auth_tkt=False):
+    """
+    By default, use a more compatible
```
"use a more compatible" what?

gstrauss replied:
apparently, that's from https://github.com/plone/plone.session/blob/master/plone/session/tktauth.py

gstrauss commented on demo/login.cgi:
```diff
@@ -0,0 +1,105 @@
+    # The original code was using the hex-string while mod_authn_tkt uses the bytes
+    digest0 = hashlib.md5(data1 + secret + data2).digest()
+    if not six.PY2:
+        # In Python 3 hashlib.md5(value).hexdigest() wants a bites value
```
s/bites/bytes/

gstrauss replied:
apparently `bites` is also from https://github.com/plone/plone.session/blob/master/plone/session/tktauth.py

gstrauss commented:
Thank you @sebas88 for your contribution!
